### ISSUE TITLE
Fine-tuning & pre-whitening

### ISSUE THREAD

**Issue by cbasavaraj** [timesler/facenet-pytorch#44]
Thanks @timesler for the great work! I was trying to fine-tune on a small dataset and noticed that the network overfits very quickly. I think it's useful to train only the last linear layer ("logits" in your code). This worked fine for me:

```
optimizer = optim.Adam(net.module.logits.parameters())
```

Unrelated-ly, the pre-whiten transform may be screwing up the identification a bit as it normalizes each face using its own mean & std-dev. In one of my test videos, an African guy is identified as a white guy. So using a standard pre-processing for all faces may be a good idea. I see there's a related pull-request. Maybe I will experiment a bit myself next week.

Thanks again.

**Comment by timesler** [comment]
@cbasavaraj thanks for using the package!

Yep, you are right that it's often better to just train the last layer. I chose not to build that into the class explicitly since there are well-known ways of doing it with base pytorch (as you've done yourself). I also didn't want to prescribe much of how the use the models.

Regarding, the prewhitening, I've now merged the change to use fixed image standardization instead, which will hopefully fix the issue you were seeing.

**Comment by maxxxpavlov** [comment]
Could you please give me examples how you fine-tune and save your model?

**Comment by cbasavaraj** [comment]
@maxxxpavlov  busy today, will get back to you tomorrow morning.

**Comment by cbasavaraj** [comment]
Hi @maxxxpavlov, Tim already has finetune.ipynb in folder examples which shows how to do fine-tuning. The only changes I have made: 

a) use dataparallel to take advantage of multiple GPUs when available:

```
CUDA = torch.cuda.is_available()
nGPU = torch.cuda.device_count()

    if CUDA:
        net = net.cuda()
        if nGPU > 1:
            net = nn.DataParallel(net)
```

b) train only the last linear layer (otherwise InceptionResNet is a deep network which overfits very fast on small datasets)
```
    logits = net.module.logits.parameters() if nGPU > 1 else net.logits.parameters()

    optimizer = optim.Adam(logits)
    scheduler = MultiStepLR(optimizer, [5, 10])
```

c) minor change on data loader, it's useful to specify num workers etc:
```
    loader_train = DataLoader(data_train, batch_size=args.batch_size, shuffle=True,
                              num_workers=nGPU*4, pin_memory=True)

    loader_valid = DataLoader(data_valid, batch_size=args.batch_size * 2,
                              num_workers=nGPU*4, pin_memory=True)
```

d) save checkpoints if validation accuracy is higher than the previous best:
```
        net.eval()
        _, vmets = training.run_epoch(
                        net, loss, loader_valid,
                        batch_metrics=metrics, show_running=True,
                        cuda=CUDA, writer=writer)

        if vmets['acc'] > accu:
            accu = vmets['acc']
            ckpt = "face_irnet_{:02d}.pth".format(epoch + 1)
            ckpt = os.path.join("./ckpts", ckpt)
            print("Saving new best model to {:}.".format(ckpt))
            torch.save({
                'epoch': epoch,
                'model': net.module.state_dict() if nGPU > 1 else net.state_dict(),
                'optim': optimizer.state_dict(),
                'vaccu': accu
                }, ckpt)
```

**Comment by elon-trump** [comment]
Can you elaborate this in detail please? where and what change we need to make in [Finetuning file](https://github.com/timesler/facenet-pytorch/blob/master/examples/finetune.ipynb).
Thanks in advance
